dbms/src/DataStreams/LimitByBlockInputStream.cpp

```diff
@@ -0,0 +1,86 @@
+#include <DB/DataStreams/LimitByBlockInputStream.h>
+
+namespace DB
+{
+
+LimitByBlockInputStream::LimitByBlockInputStream(BlockInputStreamPtr input_, size_t value_, Names columns_)
+	: columns_names(columns_)
+	, value(value_)
+{
+	children.push_back(input_);
+}
+
+String LimitByBlockInputStream::getID() const
+{
+	std::stringstream res;
+	res << "LimitBy(" << children.back()->getID() << ")";
+	return res.str();
+}
+
+Block LimitByBlockInputStream::readImpl()
+{
+	while (true)
+	{
+		Block block = children[0]->read();
+		if (!block)
+			return Block();
+
+		const ConstColumnPlainPtrs column_ptrs(getKeyColumns(block));
+		const size_t rows = block.rows();
+		const size_t old_set_size = set.size();
+		IColumn::Filter filter(rows);
+
+		for (size_t i = 0; i < rows; ++i)
+		{
+			UInt128 key;
+			SipHash hash;
+
+			for (size_t j = 0; j < column_ptrs.size(); ++j)
```

Comment:
Range based loop with be simpler. It also could be more optimal, if compiler could not decide, that `column_ptrs.size()` is loop invariant.